### PR TITLE
Initial, sandboxed Chrome Flatpak

### PR DESCRIPTION
**Pull request by refi64**
**Please put this on the beta branch!** (Also, yes this is fully sandboxed.)

It uses the new (experimental-ish) Zypak functionality that I have yet to rebase (but that can happen later). It'll run on systems with suid bwrap using the classic Zypak implementation, but after printing a warning since Chrome spawns many, many, many subprocesses, and having them properly fork off of the zygote is the best way to go there.

Requries 1.8.0+ because it needs either the new SpawnStarted signal or [this bugfix](https://github.com/flatpak/flatpak/issues/3029) to be usable. Doesn't pass `--device=all` atm due to [this bug](https://github.com/flatpak/flatpak/issues/3722).

NaCl's sandbox doesn't work, this isn't really something I'll be fixing since it's being nuked from Chrome on Linux this year anyway. To prevent it from running unsandboxed, I just removed the binaries, so it won't load at all.

**Please don't wire this to extra-data-checker for now** due to merge annoyances, I'll be updating it manually while in the beta state.

Feel free to try and break everything :rofl:

### CODE REVIEW

**Comment by tinywrkb** [comment]
libsecret is missing. With libsecret, Chrome's safe storage key is kept in the host's login keyring but passwords and cookies seem to be kept in a seperate xdg-data/keyrings/default.keyring.
```
- --talk-name=org.freedesktop.secrets
...
      - name: libsecret
        buildsystem: meson
        config-opts:
          - -Dvapi=false
          - -Dmanpage=false
          - -Dgtk_doc=false
        sources:
          - type: archive
            url: https://gitlab.gnome.org/GNOME/libsecret/-/archive/0.20.3/libsecret-0.20.3.tar.gz
            sha256: 9a90f733d4cac1c5a937af1fc44ddfc52458a4bcb015891247ca75037f446f6d
        cleanup:
          - /bin
          - /include
          - /lib/girepository-1.0
          - /share/gir-1.0
```

[There's a Chrome svg icon at Wikipedia](https://en.wikipedia.org/wiki/File:Google_Chrome_icon_(September_2014).svg).

**Comment by tinywrkb** [comment]
Please consider adding Arch Linux like `chrome-flags.conf` support, [see example](https://github.com/tinywrkb/flatpaks/blob/master/com.google.chrome/resources/com.google.chrome).

**Comment by nedrichards** [comment]
Have you asked upstream about this?

**Comment by refi64** [comment]
@tinywrkb good catch on libsecret! I used the shared-modules version and confirmed that the password database used the v11 (encrypted using a libsecret key) format to be sure. Also added Arch's chrome-flags.conf support. That being said, I'll probably stick with the current icons for now, since if the icon were to change again, it would be far easier to grab it from the official packages as done now rather than trying to extract it from an APK or hope someone does it and posts it on Wikimedia first.

**Comment by refi64** [comment]
@nedrichards Yes, the response was that supported another sandbox was something the team wasn't looking into for now.

**Comment by eszlari** [comment]
When would you move it from beta to stable?

**Comment by bilelmoussaoui** [comment]
bot, build com.google.Chrome

File: com.google.Chrome.metainfo.xml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+<!-- Copyright 2017 the Chromium authors, with modifications -->
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+  <id>com.google.Chrome</id>
+  <launchable type="desktop-id">com.google.Chrome.desktop</launchable>
+  <name>Google Chrome</name>
+  <developer_name>Google</developer_name>
+  <summary>The web browser from Google</summary>
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license>Proprietary</project_license>
```
Comment (AsciiWolf) [inline com.google.Chrome.metainfo.xml:10]: This should be `LicenseRef-proprietary`.

**Comment by refi64** [comment]
bot, build com.google.Chrome

**Comment by refi64** [comment]
@eszlari the main reason I'm proposing this for beta is just because this came with a lot of Zypak changes, thus there's a risk for some occasional bugs that need to be ironed out, thus I'm not comfortable with pushing this to stable until enough people have tried it to know it's actually, well, stable.

File: apply_extra.sh [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/sh
```
Comment (eszlari) [inline apply_extra.sh:1]: Any reason for not putting this in the manifest?

File: com.google.Chrome.yaml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  - '--share=ipc'
+  - '--share=network'
+  - '--socket=pulseaudio'
+  - '--socket=x11'
+  - '--require-version=1.8.0'
+  - '--talk-name=org.freedesktop.Notifications'
+  - '--talk-name=org.freedesktop.secrets'
+  - '--talk-name=org.freedesktop.UPower'
+  - '--talk-name=org.gnome.SessionManager'
+  - '--filesystem=home'
```
Comment (eszlari) [inline com.google.Chrome.yaml:21]: Could you add a comment, explaining why this is needed?

File: com.google.Chrome.yaml [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  - shared-modules/libsecret/libsecret.json
+
+  - name: zypak
+    sources:
+      - type: git
+        url: https://github.com/refi64/zypak
+        branch: spawn-strategy
+        commit: 0c58167d2feb1e8065eafa3df06e7f63043a8a15
+      # - type: dir
+      #   path: ../../code/zypak
```
Comment (eszlari) [inline com.google.Chrome.yaml:32]: I guess you will remove that later?!

File: apply_extra.sh (hunk 2 above)
Comment (refi64) [reply apply_extra.sh:1]: Mostly just because there's a quite decent chance it'll grow over time, so it's easier to just pull it out now.

File: com.google.Chrome.yaml (hunk 3 above)
Comment (refi64) [reply com.google.Chrome.yaml:21]: It's just the same reason as all the Electron Flatpaks, Chromium doesn't support the XDG portals.

File: com.google.Chrome.yaml (hunk 4 above)
Comment (refi64) [reply com.google.Chrome.yaml:32]: Yup! It's just easier here for now while I work on them both.

File: com.google.Chrome.yaml (hunk 3 above)
Comment (eszlari) [reply com.google.Chrome.yaml:21]: Why not just `--filesystem=xdg-download` instead?

**Comment by sr229** [comment]
I suggest shipping a Chromium build for testing before you work on Chrome itself since they're the same codebase technically, the debug symbols helps alot.

File: first_run.html [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+</head>
+
+<body>
+  <div id="content">
+    <h2>Chrome Flatpak Beta</h2>
+
+    <p>Please be aware of the following as you use this:</p>
+
+    <ul>
+      <li><b>Do NOT immediately file bugs on the Chromium bug tracker!</b> File them to the corresponding
```
Comment (sr229) [inline first_run.html:24]: I don't think you would file bugs from Chrome in the Chromium bug tracker, we have a dedicated support for Chrome itself AFAIK.

**Comment by refi64** [comment]
> I suggest shipping a Chromium build for testing before you work on Chrome itself since they're the same codebase technically, the debug symbols helps alot.

Well, the main issue here was that, turns out, most prebuilt Chromium binaries for Linux *don't* have debuginfo available. If we build it ourselves, that comes with infra requirements and such, and that might as well be using the custom Flatpak patches anyway.

TLDR I did use a local debug build for debugging, but that's not really fit for release without having a lot of other stuff first.

File: com.google.Chrome.yaml (hunk 3 above)
Comment (refi64) [reply com.google.Chrome.yaml:21]: Hmm, the main idea I had was just that you're generally a lot more likely to download to locations like the Documents folder from your browser than e.g. Discord. That being said, I could instead add the major XDG directories (like documents) here instead of the full home, would that work better?

File: first_run.html (hunk 5 above)
Comment (refi64) [reply first_run.html:24]: Oh I've definitely seen several people file bugs on the Chromium tracker without using the feedback option :sweat_smile: That being said I should definitely advise against using the feedback option as well for the moment.

**Comment by refi64** [comment]
Okay so I clarified the bug tracker notes more & locked down FS access to XDG dirs, controllers and U2F keys should work if they're plugged in *prior* to starting Chrome w/ a Flatpak 1.8.1 requirement.

File: com.google.Chrome.yaml (hunk 3 above)
Comment (Garbee) [reply com.google.Chrome.yaml]: Having the filesystem open is very useful. In fact only the home folder might be too restrictive depending on people's file setups. There are things like the new [Native Filesystem API](https://web.dev/native-file-system/) which users can select whole folders they have access to for say, editing them online.

At *least* all of home should be allowed. However, we may also want to expand this to full access later on should users come forward needing access to things outside of that space.

File: first_run.html (hunk 5 above)
Comment (Garbee) [reply first_run.html]: FYI the feedback option in Chrome just goes to the Chromium tracker. There is no "dedicated space" for one or the other.

**Comment by tinywrkb** [comment]
Chrome natively supports MPRIS2. The following works for me, though I only tested it with my own `com.google.chrome` package so maybe the D-Bus interface name below needs be to capitalize as `Chrome`.
```
finish-args:
...
- --own-name=org.mpris.MediaPlayer2.chrome.*
- --talk-name='org.mpris.MediaPlayer2.*
```

File: first_run.html (hunk 5 above)
Comment (sr229) [reply first_run.html]: @Garbee we actually do, but it's for Enterprise concerns, we usually use the Google Help Center for that.

Comment (Garbee) [reply first_run.html]: Oh interesting. Perhaps what I've seen in the main hose previously was dumps from that when not necessary or just an old method of managing things. Good to know there is some separation there in case it comes up in the future.

**Comment by refi64** [comment]
@tinywrkb interesting, I admittedly didn't realize that was a thing. Just added, did some tests and indeed --talk-name isn't needed.

**Comment by bilelmoussaoui** [comment]
bot, build com.google.Chrome

**Comment by TheEvilSkeleton** [comment]
Not sure if anybody is interested, but I tried it and it's working very well right now in Fedora!

I used [gasinvein/flatpak-edge](https://copr.fedorainfracloud.org/coprs/gasinvein/flatpak-edge)'s repo to get the latest Flatpak release. The latest in the official Fedora repos were using was 1.6.4 (I think), which would give me this error:

```
Error: com.google.Chrome needs a later flatpak version
error: Failed to install com.google.Chrome: app/com.google.Chrome/x86_64/test needs a later flatpak version (1.8.1)
```

So, I tried:

Test | Working
-- | --
Load YouTube | Yes
Played a video | Yes
Played a video in 60fps | Yes
Picture-in-picture | Yes
Logged into Google | Yes
Installed extensions | Yes
Logged into Facebook | Yes
Restricted filesystem subsets. Revoked access to `xdg-videos`, `xdg-music` and `xdg-documents` | Yes
Downloaded files | Yes
Tampered in chrome://flags | Yes

So far it's working pretty well, Amazing work you have done @refi64!

**Comment by Erick555** [comment]
@TheEvilSkeleton could you test h264 playback? For example by installing [this extension](https://chrome.google.com/webstore/detail/h264ify/aleakchihdccplidncghkekgioiakgal?hl=en) and browse some videos on youtube. There were problems with h264 playback on firefox flatpak so I wonder if chrome is affected too.

**Comment by barthalion** [comment]
I'm pretty sure Chrome just ships own ffmpeg libraries.

**Comment by tinywrkb** [comment]
I'm posting this even though I haven't made the minimal effort to figure out if something is not set up correctly on my end or debug this further. Anyway, the two issues I've noticed are:

1. Restart is broken (chrome://restart). Probably it doesn't matter but I'm pretty sure it worked correctly with zypak v2020.02.  
The error message is
`bwrap: execvp /app/bin/../lib/zypak-helper: No such file or directory`
2. flatpak-portal crashes when Chrome is the first Flatpak app launched after reboot, and of course Chrome will not load. Chrome will start up correctly if launched again. I'm not sure what's going on, and I didn't try looking further at this.  
Note that I have a minimal host system with a windows manager (i3wm) and the only packages that depend on a widget toolkit are gnome-secret/gcr, xdg-desktop-portal-gtk, and xdg-desktop-portal.  
So there's nothing there to start flatpak-portal before launching Chrome, so maybe there's a race.  
I guess I can maybe work around this by starting flatpak-portal.service after boot and not waiting for dbus to do this when Chrome is launched, but this needs proper debugging.

Other things I noticed but need confirmation or testing:

* USB security tokens do not work, or at least my Yubikey didn't work. I have everything set up correctly on my host and it worked before when I was using Chromium installation on the host system.
* Chrome trying to find the system bus socket but after looking at dbus-monitor output (with system socket override) it doesn't seem like Chrome actually tries to send messages.
* I don't use this feature but Chrome NSS code does not support XDG directories standard spec, so maybe it will be a good idea to bind mount ~/.pki/nssdb so the user can gain access to certs from the host. There's a [relevant bug report here](https://bugs.chromium.org/p/chromium/issues/detail?id=1038587). Last time I checked the relevant code was in Chromium's `crypto/nss_util.cc`.

**Comment by TheEvilSkeleton** [comment]
> @TheEvilSkeleton could you test h264 playback? For example by installing [this extension](https://chrome.google.com/webstore/detail/h264ify/aleakchihdccplidncghkekgioiakgal?hl=en) and browse some videos on youtube. There were problems with h264 playback on firefox flatpak so I wonder if chrome is affected too.

I installed the extension, and here are all the videos I tried:

- https://youtu.be/TvFghprIeMI
- https://youtu.be/e8QtsyNXvFg
- https://youtu.be/iEQu3FzGyfQ
- https://youtu.be/gUeT0nIqE88
- https://youtu.be/Z91g1gHt9gU

All those videos worked well :)

**Comment by refi64** [comment]
Chrome itself ships the H.264 support and the Flatpak doesn't touch it, so that should definitely work consistently. (Of course, do note upstream Chrome doesn't support stuff like VA-API, thus neither will this Flatpak.)

File: com.google.Chrome.yaml (hunk 3 above)
Comment (refi64) [reply com.google.Chrome.yaml]: @Garbee (missed this comment previously, sorry) I think it would be easier to expand the set of directories later on in the beta pipeline than restrict it. Recently I was looking through the source, and I think that there *might* actually be a way to force it to use portals, but I'm not sure how that would interact with the native filesystem API; at a glance, it would need probably some more integrated support to regain access to the files after app restart.

At minimum, for the immediate moment, the APIs are still in origin trial, so I think we could get away with a locked down filesystem view at least for the beta phase while the details get sorted out.

File: chrome.sh [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  touch "$run_stamp"
+fi
+
+if [[ ! -f "$mimic_stamp" ]] && ! zypak-helper spawn-strategy-test; then
+  zenity --info --title='Chrome Flatpak' --no-wrap \
+    --text="$(< /app/share/flatpak-chrome/mimic_warning.txt)"
+  touch "$mimic_stamp"
+fi
+
+if [[ -f "$XDG_CONFIG_HOME/chrome-flags.conf" ]]; then
```
Comment (Erick555) [inline chrome.sh:33]: I think on ubuntu additional flags are set in `~/.chromium-browser.init`

**Comment by randoogle** [comment]
> Build [23771 successful](https://flathub.org/builds/#/builders/32/builds/23771)
> To test this build, install it from the testing repository:
> 
> ```
> flatpak install --user https://dl.flathub.org/build-repo/22839/com.google.Chrome.flatpakref
> ```
Is there something wrong with the servers, or did you remove this file? I wanted to try it out, but I get a 404 response when trying the url you mentioned.

> error: Can't load uri https://dl.flathub.org/build-repo/22839/com.google.Chrome.flatpakref: Server returned status 404: Not Found

**Comment by Erick555** [comment]
Test builds are automatically removed after few days.

**Comment by LeandroStanger** [comment]
bot, build com.google.Chrome

**Comment by tinywrkb** [comment]
@randoogle 

```
git clone https://github.com/flathub/flathub.git && cd flathub
git fetch origin pull/1630/head:chrome
git checkout chrome
git submodule update --init
export FPB_WRKDIR=/tmp/wrk # or somewhere else
mkdir -p $FPB_WRKDIR
flatpak-builder \
  --install \
  --user \
  --force-clean \
  --state-dir=${FPB_WRKDIR}/flatpak-builder \
  --repo=${FPB_WRKDIR}/flatpak-repo  \
  ${FPB_WRKDIR}/flatpak-target \
  com.google.Chrome.yaml
```

**Comment by sr229** [comment]
> I'm pretty sure Chrome just ships own ffmpeg libraries.

Chromium does support such since it doesn't rely in the host-side ffmpeg IIRC

**Comment by sr229** [comment]
> I'm pretty sure Chrome just ships own ffmpeg libraries.

Chromium does support such since it doesn't rely in the host-side ffmpeg.

File: com.google.Chrome.yaml (hunk 3 above)
Comment (sr229) [reply com.google.Chrome.yaml]: I would suggest having the home filesystem open for the user since the Native Filesystem API might already be in use and this would cause a lot of finnicky errors.

**Comment by randoogle** [comment]
@tinywrkb 
Thanks for the build info. I ended up getting this error near the end of the build, though:

> error: Unknown filesystem location host-etc, valid locations are: host, home, xdg-*[/…], ~/dir, /dir
Error: Child process exited with code 1

Running Fedora Silverblue 32, if that matters.

**Comment by tinywrkb** [comment]
@randoogle the minimal requirement for this app is Flatpak v1.8.1. There's no point in changing the yaml manifest, the app will not work correctly without the fixes that are included in the latest Flatpak release.

**Comment by refi64** [comment]
@randoogle fwiw I've been using 1.8.1 on Silverblue, just by grabbing the RPMs off Koji and using `rpm-ostree override` (which isn't particularly pretty, but hey, it works :sweat_smile:), no issues so far.

File: com.google.Chrome.yaml (hunk 3 above)
Comment (refi64) [reply com.google.Chrome.yaml]: Filesystem access seems to be more debated than I thought :sweat_smile: at minimum like I said, we can probably get away with using this for the beta, and commit to deciding which direction to go in (or maybe if I can make Zypak override to use the native file chooser :eyes:) by an eventual stable release. Native filesystem API is still in origin trials, I don't anticipate this taking *that* long to determine.

**Comment by refi64** [comment]
Any particular hold ups remaining?

File: com.google.Chrome.yaml [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      - |
+          for icon in 16 24 32 48 64 128 256; do
+            install -Dm 644 com.google.Chrome-$icon.png /app/share/icons/hicolor/${icon}x${icon}/apps/com.google.Chrome.png
+          done
+
+      # For extra-data to be able to extract the .deb
+      - 'install -Dm 755 /{usr,app}/bin/ar'
+      - 'cp /usr/lib/$(gcc --print-multiarch)/libbfd-*.so /app/lib'
+    sources:
+      - type: extra-data
```
Comment (bilelmoussaoui) [inline com.google.Chrome.yaml:59]: This needs the flathub-data-checker thing

File: com.google.Chrome.yaml [hunk 8]
```diff
@@ -0,0 +1,99 @@
+app-id: com.google.Chrome
+desktop-file-name-prefix: '(Testing) '
```
Comment (bilelmoussaoui) [inline com.google.Chrome.yaml:2]: Is this needed? we can otherwise just push to beta instead of stable directly

File: com.google.Chrome.yaml (hunk 7 above)
Comment (nedrichards) [reply com.google.Chrome.yaml:59]: Per the message in the top comment:
> Please don't wire this to extra-data-checker for now due to merge annoyances, I'll be updating it manually while in the beta state.

I'm OK with merging this to beta without it but *not* pushing to stable without it as web browsers as the largest opportunity for most desktop Linux users to evaluate untrusted content (with the issues that come along with that).

**Comment by nedrichards** [comment]
bot, build com.google.Chrome

File: com.google.Chrome.yaml (hunk 8 above)
Comment (refi64) [reply com.google.Chrome.yaml:2]: Is the prefix automatically added for beta apps? (The intent was to push to beta regardless, which I mentioned in the main PR message.)

File: com.google.Chrome.yaml (hunk 7 above)
Comment (refi64) [reply com.google.Chrome.yaml:59]: Yeah I *definitely* wouldn't want this on stable without the external data checker set up yet.

**Comment by nedrichards** [comment]
OK, I think this is good enough for beta testing but, just like firefox I don't anticipate a terribly quick transition to stable as there's a fair few rough edges to knock off and I'd want the >=1.8.1 graph [on here](https://ahayzen.com/direct/flathub.html) to be a *lot* higher before pushing such a high profile app.

**Comment by FilBot3** [comment]
So, how does one build this project? It doesn't appear in Flathub either.

**Comment by cho2** [comment]
> So, how does one build this project? It doesn't appear in Flathub either.

It's on `flathub-beta`
